plugins/chain_plugin/trx_retry_db.cpp

```diff
@@ -0,0 +1,326 @@
+   size_t memory_size()const { return ptrx->get_estimated_size() * 3 + sizeof(*this); }
+
+   tracked_transaction(const tracked_transaction&) = delete;
+   tracked_transaction() = delete;
+   tracked_transaction& operator=(const tracked_transaction&) = delete;
+   tracked_transaction(tracked_transaction&&) = default;
+};
+
+struct by_trx_id;
+struct by_expiry;
+struct by_ready_block_num;
+struct by_block_num;
+struct by_last_try;
+
+using tracked_transaction_index_t = multi_index_container<tracked_transaction,
+      indexed_by<
+            hashed_unique<tag<by_trx_id>,
+                  const_mem_fun<tracked_transaction, const transaction_id_type&, &tracked_transaction::id>, std::hash<transaction_id_type>
+            >,
+            ordered_non_unique<tag<by_expiry>,
+                  const_mem_fun<tracked_transaction, fc::time_point_sec, &tracked_transaction::expiry>
+            >,
+            ordered_non_unique<tag<by_ready_block_num>,
+                  const_mem_fun<tracked_transaction, uint32_t, &tracked_transaction::ready_block_num>
+            >,
+            ordered_non_unique<tag<by_block_num>,
+                  member<tracked_transaction, uint32_t, &tracked_transaction::block_num>
+            >,
+            ordered_non_unique<tag<by_last_try>,
+                  member<tracked_transaction, fc::time_point, &tracked_transaction::last_try>
+            >
+      >
+>;
+
+} // anonymous namespace
+
+namespace eosio::chain_apis {
+
+struct trx_retry_db_impl {
+   explicit trx_retry_db_impl(const chain::controller& controller, size_t max_mem_usage_size,
+                              fc::microseconds retry_interval, fc::microseconds max_expiration_time,
+                              fc::microseconds abi_serializer_max_time)
+   : _controller(controller)
+   , _transaction_ack_channel(appbase::app().get_channel<chain::plugin_interface::compat::channels::transaction_ack>())
+   , _abi_serializer_max_time(abi_serializer_max_time)
+   , _max_mem_usage_size(max_mem_usage_size)
+   , _retry_interval(retry_interval)
+   , _max_expiration_time(max_expiration_time)
+   {}
+
+   const fc::microseconds& get_max_expiration_time()const {
+      return _max_expiration_time;
+   }
+
+   size_t size()const {
+      return _tracked_trxs.index().size();
+   }
+
+   void track_transaction( packed_transaction_ptr ptrx, std::optional<uint16_t> num_blocks, next_function<std::unique_ptr<fc::variant>> next ) {
+      EOS_ASSERT( _tracked_trxs.memory_size() < _max_mem_usage_size, tx_resource_exhaustion,
+                  "Transaction exceeded  transaction-retry-max-storage-size-gb limit: ${m} bytes", ("m", _tracked_trxs.memory_size()) );
+      auto i = _tracked_trxs.index().get<by_trx_id>().find( ptrx->id() );
+      if( i == _tracked_trxs.index().end() ) {
+         _tracked_trxs.insert( {std::move(ptrx), !num_blocks.has_value() ? lib_totem : *num_blocks, 0, {}, fc::time_point::now(), next} );
```

Comment:
Will this work correctly if user indicates num-blocks = 0?

Reply:
It should, and it is a valid use case. A trx comes in early during speculative execution, is relayed to a bp which puts it in its current block and then sends you the block. You abort your block and process that trx in that block just received, you would then ack it instead of waiting for another block. Added a test case for it to the unittest.